### ISSUE TITLE
Managed endpoint doesn't use Azure ML connections (private pip) when building image

### ISSUE THREAD

**Issue by thomasfrederikhoeck** [Azure/azure-sdk-for-python#22441]
- **Package Name**: azureml-mlflow
- **Package Version**: 1.37.0
- **Operating System**: Windows
- **Python Version**: 3.8.11

**Describe the bug**
When using private pip feeds it is possible to set up a connection (token) which is shared across the Azure ML Studio. This works for creating an environment for training the model. If the model is saved in the MLFLOW format it also works when using an Azure Container Instance for creating a realtime endpoint very easily. But when trying to use the managed endpoint the token is not used.

**To Reproduce**

When using private pip feeds it is possible to set up a connection (token) which is shared across the Azure ML Studio by running
1:
```python
ws.set_connection(
    name="private-pip",
    category= "PythonFeed",
    authType="PAT",
    target = "https://@pkgs.dev.azure.com/MY-ORG",
    value = "1234abcd"
)
```

2: This will then be used if you construct a environment were a private pip is referred and replace the package `myprivatepackage` with something in you feed:
```python
from azureml.core import Environment
from azureml.core.conda_dependencies import CondaDependencies

env = Environment("test_env")
cd = CondaDependencies.create(
    python_version="3.8.12",
    conda_packages=["pip"],
    pip_packages=["scikit-learn", "pandas","azureml-mlflow","pyyaml","myprivatepackage"]
    )
cd.set_pip_option("--index-url https://pkgs.dev.azure.com/MY-ORG/_packaging/MY-FEED/pypi/simple/")
env.python.conda_dependencies = cd
env.register(ws)
env.build(ws)
```
3: If you train a model you can deploy it where it connects using the token when building the image too. 

First create a training file:
```python
## ml_project/train.py
from sklearn.linear_model import LinearRegression
import mlflow
import mlflow.sklearn
from azureml.core import Run
import yaml

from sklearn.datasets import load_iris
X, y = load_iris(as_frame=True, return_X_y=True)

run = Run.get_context()
ws = run.experiment.workspace
env = run.get_environment()
conda_env = yaml.safe_load(env.python.conda_dependencies.serialize_to_string())

with mlflow.start_run():
    X, y = load_iris(as_frame=True, return_X_y=True)
    model = LinearRegression()
    model.fit(X, y)
    mlflow.sklearn.log_model(model, "model", conda_env=conda_env)


```
Then then train the model
```python

from azureml.core import Experiment, ScriptRunConfig
from azureml.core.compute import ComputeTarget

cpu_cluster = ComputeTarget(workspace=ws, name="test-cluster")
exp = Experiment(workspace=ws, name="test")

src = ScriptRunConfig(
    source_directory= "mlproject",
    script='train.py',
    compute_target=cpu_cluster,
    environment=env
)
run = exp.submit(src)
```

Then register the model:
```python
import mlflow
mlflow.set_tracking_uri(ws.get_mlflow_tracking_uri())
model_info = mlflow.register_model(f"runs:/{run.id}/model", "test_model")
```

4: You can deploy it using Azure Container Instance which works by grabbing the token. It deploys after some time.:
```python
from mlflow.deployments import get_deploy_client
mlflow.set_tracking_uri(ws.get_mlflow_tracking_uri())
client = get_deploy_client(mlflow.get_tracking_uri())
client.create_deployment(model_uri=f'models:/test_model/1', name="mlflow-test-aci")
```
5: Go to the UI for Azure Machine Learning Studio and go to the wizard for setting up a managed endpoint. As the model is of type MLFLOW it will create enviorment and scoring script for you. When the image builds it will fail as it is expecting the user to put in a password:
```
Instance status:
SystemSetup: Succeeded
UserContainerImagePull: Succeeded
ModelDownload: Succeeded
UserContainerStart: InProgress

Container logs:
2022-01-11T12:29:04,946980587+00:00 - rsyslog/run 
2022-01-11T12:29:04,951881106+00:00 - gunicorn/run 

2022-01-11T12:29:04,953826713+00:00 - nginx/run 
AzureML image information: mlflow-ubuntu18.04-py37-cpu-inference:20210930.v4


PATH environment variable: /opt/miniconda/envs/amlenv/bin:/opt/miniconda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
PYTHONPATH environment variable: 

Pip Dependencies
---------------
applicationinsights==0.11.10
azureml-inference-server-http==0.3.2
certifi==2021.5.30
charset-normalizer==2.0.6
click==8.0.1
cloudpickle==2.0.0
databricks-cli==0.15.0
entrypoints==0.3
Flask==1.0.3
gitdb==4.0.7
GitPython==3.1.24
gunicorn==20.1.0
idna==3.2
importlib-metadata==4.8.1
inference-schema==1.3.0
itsdangerous==2.0.1
Jinja2==3.0.1
MarkupSafe==2.0.1
mlflow-skinny==1.17.0
numpy==1.21.2
pandas==1.2.4
protobuf==3.18.0
python-dateutil==2.8.2
pytz==2021.1
PyYAML==5.4.1
requests==2.26.0
six==1.16.0
smmap==4.0.0
tabulate==0.8.9
typing-extensions==3.10.0.2
urllib3==1.26.7
Werkzeug==2.0.1
wrapt==1.12.1
zipp==3.6.0


Entry script directory: /var/mlflow_resources/.

Updating conda environment from /var/azureml-app/azureml-models/test_model/1/model/conda.yaml !
Collecting package metadata (repodata.json): ...working... done
Solving environment: ...working... done

Downloading and Extracting Packages

ca-certificates-2020 | 128 KB    |            |   0% 
ca-certificates-2020 | 128 KB    | #2         |  13% 
ca-certificates-2020 | 128 KB    | ########## | 100% 

_openmp_mutex-4.5    | 22 KB     |            |   0% 
_openmp_mutex-4.5    | 22 KB     | ########## | 100% 

libgcc-ng-11.2.0     | 887 KB    |            |   0% 
libgcc-ng-11.2.0     | 887 KB    | ########## | 100% 
libgcc-ng-11.2.0     | 887 KB    | ########## | 100% 

libzlib-1.2.11       | 59 KB     |            |   0% 
libzlib-1.2.11       | 59 KB     | ########## | 100% 

certifi-2020.6.20    | 160 KB    |            |   0% 
certifi-2020.6.20    | 160 KB    | ########## | 100% 

ld_impl_linux-64-2.3 | 667 KB    |            |   0% 
ld_impl_linux-64-2.3 | 667 KB    | ########## | 100% 
ld_impl_linux-64-2.3 | 667 KB    | ########## | 100% 

libnsl-2.0.0         | 31 KB     |            |   0% 
libnsl-2.0.0         | 31 KB     | ########## | 100% 

sqlite-3.37.0        | 1.5 MB    |            |   0% 
sqlite-3.37.0        | 1.5 MB    | ########## | 100% 
sqlite-3.37.0        | 1.5 MB    | ########## | 100% 

xz-5.2.5             | 438 KB    |            |   0% 
xz-5.2.5             | 438 KB    | ########## | 100% 
xz-5.2.5             | 438 KB    | ########## | 100% 

libstdcxx-ng-11.2.0  | 4.2 MB    |            |   0% 
libstdcxx-ng-11.2.0  | 4.2 MB    | ########## | 100% 
libstdcxx-ng-11.2.0  | 4.2 MB    | ########## | 100% 

pip-20.2.4           | 2.0 MB    |            |   0% 
pip-20.2.4           | 2.0 MB    | ########## | 100% 
pip-20.2.4           | 2.0 MB    | ########## | 100% 

readline-8.1         | 295 KB    |            |   0% 
readline-8.1         | 295 KB    | ########## | 100% 

libgomp-11.2.0       | 427 KB    |            |   0% 
libgomp-11.2.0       | 427 KB    | ########## | 100% 
libgomp-11.2.0       | 427 KB    | ########## | 100% 

ncurses-6.2          | 1.1 MB    |            |   0% 
ncurses-6.2          | 1.1 MB    | ########## | 100% 
ncurses-6.2          | 1.1 MB    | ########## | 100% 

python-3.8.12        | 26.2 MB   |            |   0% 
python-3.8.12        | 26.2 MB   | 5          |   6% 
python-3.8.12        | 26.2 MB   | ##9        |  30% 
python-3.8.12        | 26.2 MB   | ####6      |  46% 
python-3.8.12        | 26.2 MB   | ######9    |  70% 
python-3.8.12        | 26.2 MB   | ########## | 100% 
python-3.8.12        | 26.2 MB   | ########## | 100% 

wheel-0.35.1         | 36 KB     |            |   0% 
wheel-0.35.1         | 36 KB     | ########## | 100% 

_libgcc_mutex-0.1    | 3 KB      |            |   0% 
_libgcc_mutex-0.1    | 3 KB      | ########## | 100% 

libffi-3.4.2         | 57 KB     |            |   0% 
libffi-3.4.2         | 57 KB     | ########## | 100% 

zlib-1.2.11          | 86 KB     |            |   0% 
zlib-1.2.11          | 86 KB     | ########## | 100% 

openssl-3.0.0        | 2.9 MB    |            |   0% 
openssl-3.0.0        | 2.9 MB    | ########## | 100% 
openssl-3.0.0        | 2.9 MB    | ########## | 100% 

tk-8.6.11            | 3.3 MB    |            |   0% 
tk-8.6.11            | 3.3 MB    | ########## | 100% 
tk-8.6.11            | 3.3 MB    | ########## | 100% 

setuptools-50.3.0    | 902 KB    |            |   0% 
setuptools-50.3.0    | 902 KB    | ########## | 100% 
setuptools-50.3.0    | 902 KB    | ########## | 100% 
Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done


==> WARNING: A newer version of conda exists. <==
  current version: 4.9.2
  latest version: 4.11.0

Please update conda by running

    $ conda update -n base -c defaults conda


Pip subprocess error:
ERROR: Exception:
Traceback (most recent call last):
  File "/opt/miniconda/envs/userenv/lib/python3.8/site-packages/pip/_internal/cli/base_command.py", line 228, in _main
    status = self.run(options, args)
  File "/opt/miniconda/envs/userenv/lib/python3.8/site-packages/pip/_internal/cli/req_command.py", line 182, in wrapper
    return func(self, options, args)
  File "/opt/miniconda/envs/userenv/lib/python3.8/site-packages/pip/_internal/commands/install.py", line 323, in run
    requirement_set = resolver.resolve(
  File "/opt/miniconda/envs/userenv/lib/python3.8/site-packages/pip/_internal/resolution/legacy/resolver.py", line 183, in resolve
    discovered_reqs.extend(self._resolve_one(requirement_set, req))
  File "/opt/miniconda/envs/userenv/lib/python3.8/site-packages/pip/_internal/resolution/legacy/resolver.py", line 388, in _resolve_one
    abstract_dist = self._get_abstract_dist_for(req_to_install)
  File "/opt/miniconda/envs/userenv/lib/python3.8/site-packages/pip/_internal/resolution/legacy/resolver.py", line 339, in _get_abstract_dist_for
    self._populate_link(req)
  File "/opt/miniconda/envs/userenv/lib/python3.8/site-packages/pip/_internal/resolution/legacy/resolver.py", line 305, in _populate_link
    req.link = self._find_requirement_link(req)
  File "/opt/miniconda/envs/userenv/lib/python3.8/site-packages/pip/_internal/resolution/legacy/resolver.py", line 270, in _find_requirement_link
    best_candidate = self.finder.find_requirement(req, upgrade)
  File "/opt/miniconda/envs/userenv/lib/python3.8/site-packages/pip/_internal/index/package_finder.py", line 900, in find_requirement
    best_candidate_result = self.find_best_candidate(
  File "/opt/miniconda/envs/userenv/lib/python3.8/site-packages/pip/_internal/index/package_finder.py", line 883, in find_best_candidate
    candidates = self.find_all_candidates(project_name)
  File "/opt/miniconda/envs/userenv/lib/python3.8/site-packages/pip/_internal/index/package_finder.py", line 827, in find_all_candidates
    package_links = self.process_project_url(
  File "/opt/miniconda/envs/userenv/lib/python3.8/site-packages/pip/_internal/index/package_finder.py", line 791, in process_project_url
    html_page = self._link_collector.fetch_page(project_url)
  File "/opt/miniconda/envs/userenv/lib/python3.8/site-packages/pip/_internal/index/collector.py", line 629, in fetch_page
    return _get_html_page(location, session=self.session)
  File "/opt/miniconda/envs/userenv/lib/python3.8/site-packages/pip/_internal/index/collector.py", line 441, in _get_html_page
    resp = _get_html_response(url, session=session)
  File "/opt/miniconda/envs/userenv/lib/python3.8/site-packages/pip/_internal/index/collector.py", line 138, in _get_html_response
    resp = session.get(
  File "/opt/miniconda/envs/userenv/lib/python3.8/site-packages/pip/_vendor/requests/sessions.py", line 543, in get
    return self.request('GET', url, **kwargs)
  File "/opt/miniconda/envs/userenv/lib/python3.8/site-packages/pip/_internal/network/session.py", line 421, in request
    return super(PipSession, self).request(method, url, *args, **kwargs)
  File "/opt/miniconda/envs/userenv/lib/python3.8/site-packages/pip/_vendor/requests/sessions.py", line 530, in request
    resp = self.send(prep, **send_kwargs)
  File "/opt/miniconda/envs/userenv/lib/python3.8/site-packages/pip/_vendor/requests/sessions.py", line 650, in send
    r = dispatch_hook('response', hooks, r, **kwargs)
  File "/opt/miniconda/envs/userenv/lib/python3.8/site-packages/pip/_vendor/requests/hooks.py", line 31, in dispatch_hook
    _hook_data = hook(hook_data, **kwargs)
  File "/opt/miniconda/envs/userenv/lib/python3.8/site-packages/pip/_internal/network/auth.py", line 256, in handle_401
    username, password, save = self._prompt_for_password(parsed.netloc)
  File "/opt/miniconda/envs/userenv/lib/python3.8/site-packages/pip/_internal/network/auth.py", line 226, in _prompt_for_password
    username = ask_input("User for {}: ".format(netloc))
  File "/opt/miniconda/envs/userenv/lib/python3.8/site-packages/pip/_internal/utils/misc.py", line 259, in ask_input
    return input(message)
EOFError: EOF when reading a line


CondaEnvException: Pip failed

Installing pip dependencies: ...working... Ran pip subprocess with arguments:
['/opt/miniconda/envs/userenv/bin/python', '-m', 'pip', 'install', '-U', '-r', '/tmp/condaenv.i5j0upr4.requirements.txt']
Pip subprocess output:
Looking in indexes: https://pkgs.dev.azure.com/MY-ORG/_packaging/MY-FEED/pypi/simple/
User for pkgs.dev.azure.com: 
failed
```
6: I assume the same errro happens on managed Batch endpoint but I haven't tried it out. 

**Expected behavior**
I expect the managed endpoint to work. I also expect it to work for batch endpoint.

**Suggestion**
A good solution to this would be to allow for setting a variable in Azure ML studio which was always loaded (at build time) as an env var (PIP_INDEX_URL) when building an environment. That way we don't have to specify the url in the environment/conda file and it can easily be changed if the pip feed changes it address.

**Comment by mccoyp** [comment]
Hi @thomasfrederikhoeck, thank you for opening an issue! I'll tag the appropriate folks so we can look into this as soon as possible.

**Comment by thomasfrederikhoeck** [comment]
Allowing to set `PIP_INDEX_URL` globally would also make it easier to follow best practices of using only one private feed described in this Microsoft white paper https://azure.microsoft.com/en-us/resources/3-ways-to-mitigate-risk-using-private-package-feeds/

**Comment by thomasfrederikhoeck** [comment]
Hi @mccoyp do you know if this has been picked up by anyone? It is stopping us from using managed endpoints with mlflow as our models uses internal packages :-)

**Comment by mccoyp** [comment]
Hi @thomasfrederikhoeck, thanks for the ping! @bandsina, have you been able to look into this, or do you know of someone on the ML team who can?

**Comment by thomasfrederikhoeck** [comment]
Hi again @mccoyp  . Sorry for keeping pinging you but this blocks us from using managed endpoints :-) I could seem that @bandsina has disabled notifications. Is it possible that you try another (internal) channel?

**Comment by mccoyp** [comment]
@thomasfrederikhoeck this is the place -- I'm sorry you haven't gotten any updates yet! I'll remind the ML team about this issue and re-tag it for their triage system.

**Comment by singankit** [comment]
@thomasfrederikhoeck Can you please share the conda.yml file from your MLflow model? If it doesn't contain any sensitive information. This will help us understand why AML is not able to identify that a private feed is there, and service connection needs to be used.

**Comment by thomasfrederikhoeck** [comment]
@singankit Yes, Here it is:
```yaml
channels:
- defaults
- conda-forge
dependencies:
- python=3.8.0
- pip
- pip:
  - --index-url https://pkgs.dev.azure.com/MY-ORG/_packaging/MY-FEED/pypi/simple/
  - mlflow
  - scikit-learn==0.24.1
  - mysecretpackage==1.0.0
name: mlflow-env
```

This then aligns with the connection if have defined:
```python
from azureml.core import Workspace
ws = Workspace(<input>)
ws.list_connections()
>>> [{'id': '/subscriptions/<sub-id>/resourceGroups/<rg>/providers/Microsoft.MachineLearningServices/workspaces/<workspace>/connections/my-pip', 'name': 'my-pip', 'type': 'Microsoft.MachineLearningServices/workspaces/connections', 'properties': {'category': 'PythonFeed', 'target': 'https://@pkgs.dev.azure.com/MY-ORG', 'authType': 'PAT', 'value': '<PAT>'}}]
```

**Comment by vizhur** [comment]
env.register(ws)
env.build(ws)

register returns an instance you'd want to build

regenv= env.register(ws)
regenv.build(ws)

**Comment by MrCordeiro** [comment]
Had a [very similar situation](https://github.com/MicrosoftDocs/azure-docs/issues/87920) when trying to build a Docker image that uses a private package.

**Comment by thomasfrederikhoeck** [comment]
@MrCordeiro It looks similar and might be related but the error I'm hitting is not the same. Step 1-4 works fine and I think your issues was related to step 1-2. :-) It is when trying to deploying a model of type MLFLOW using Managed endpoint.

**Comment by thomasfrederikhoeck** [comment]
Do you have any updates on this issue, @singankit ? :-) We would love to start using managed endpoints but this is holdings us back.

**Comment by amolb089** [comment]
@thomasfrederikhoeck Managed endpoint currently does not support using private feeds. The workaround for now is for customers is to BYOC(bring your own container). Build their own dockerfile with appropriate permission. Did you follow any public documentation for managed endpoint which mentions this scenario? I can take that feedback and provide it to the team to correct. Let me know if there is anything else that I can help you with!

**Comment by thomasfrederikhoeck** [comment]
@amolb089  Thank you for the answer. How can I specify my own image when using the MLFlow model format and managed endpoints? I don't think I have seen it described? I just assumed it worked in similar fashion to `https://docs.microsoft.com/en-us/azure/machine-learning/how-to-deploy-mlflow-models` which does work for private feeds :-)

**Comment by singankit** [comment]
@thomasfrederikhoeck apologizes for the delay. Since this is an unsupported scenario we are working on a custom sample which can help with this scenario. However it will require using v2 [CLI](https://docs.microsoft.com/en-us/azure/machine-learning/how-to-configure-cli?tabs=cli) to create Managed Online Deployment. Will share the example in couple of days.

**Comment by thomasfrederikhoeck** [comment]
Thank you @singankit . Will you post it here when it is available?

**Comment by thomasfrederikhoeck** [comment]
@singankit @amolb089  do you have an update on an example on how to achieve it? :-)

**Comment by rgreen1995** [comment]
@thomasfrederikhoeck 
@singankit 

Was there ever a solution to this? I'm having a similar problem when trying to switch to endpoints. I want to be able to update the env in v2 using a private wheel in the same way that you could in the v1 SDK i.e. : 

```
from azureml.core import Environment
from azureml.core.environment import CondaDependencies

conda_dependencies = CondaDependencies(
        conda_dependencies_file_path="../environment.yml"
    )
env = Environment(name="name")
conda_dependencies.set_python_version("version")
env.python.conda_dependencies = conda_dependencies
whl_url = env.add_private_pip_wheel(workspace, 
                               file_path="path_to_whl.whl",
                               exist_ok=True
  )
env.python.conda_dependencies.add_pip_package(whl_url)
```

Any help would be much appreciated

**Comment by vizhur** [comment]
private wheels are not supported in v2. though if you reference it from the storage associated with your ws AzureML will be able to resolve it, as a backward compat support. I'd recommend to use private indexes instead as described above. Other solution if your wheel is relatively small is to use docker context approach and have it there.

**Comment by rgreen1995** [comment]
@vizhur I'm not sure I follow, how do I actually add the package to azure without using v1 Environment class?, Basically the equivalent of my code above or step 2 in the initial question using Environment from v2?

**Comment by vizhur** [comment]
There is no equivalent in V2 for private packages

**Comment by thomasfrederikhoeck** [comment]
TL;DR: You can't use a private index with v2 and MLFlow unless you put your secret in clear text.

@rgreen1995 Not to my knowledge but your problem is different as we use index instead of wheel. We abandoned trying to use managed endpoints (v2 in general). 

@vizhur I'm not quite sure that i follow? This whole issue is about private indexes doesn't work for v2 (managed endpoints) unless you put the token for the index as clear text in the requirements file. It can't load the from connections as I have descibed above. I guess as a work around you can build the environment separately and but this doesn't work when using MLFlow as you can't select the enviorment (docker image) in that case.

**Comment by vizhur** [comment]
Hey @thomasfrederikhoeck ,  @rgreen1995 here https://github.com/Azure/azure-sdk-for-python/issues/22441#issuecomment-1263669644 had a question about private wheel feature, that is a bit out of the issue context.

**Comment by thomasfrederikhoeck** [comment]
@vizhur I guess :-) But do you know if there is any progress with private pip index for managed endpoints?

**Comment by thomasfrederikhoeck** [comment]
@vizhur If you don't know, can you provide who can? :-)

**Comment by vizhur** [comment]
I'd bet on @singankit ;)

**Comment by thomasfrederikhoeck** [comment]
@vizhur can I maybe convince you to ping them internally?

**Comment by polavishnu4444** [comment]
Is there any working solution or workaround for this currently to use private repos?

**Comment by thomasfrederikhoeck** [comment]
or maybe @azureml-github is better?

**Comment by rgreen1995** [comment]
+1 any documented working solution would be great, this is the only thing stopping me from transitioning to endpoints

**Comment by thomasfrederikhoeck** [comment]
From my understanding connections should still be a thing but it is not very clear if they should still be able to be used for private pip index: https://learn.microsoft.com/en-us/azure/machine-learning/how-to-migrate-from-v1#connection-workspace-connection-in-v1

**Comment by thomasfrederikhoeck** [comment]
I saw this but does this mean that it is supported? https://github.com/Azure/azureml-examples/blob/main/sdk/python/resources/connections/connections.ipynb

**Comment by luigiw** [comment]
@thomasfrederikhoeck Yes, workspace connections can be used for private feeds. And the environment you build using private feeds can be used in v2 deployments/endpoints.

**Comment by thomasfrederikhoeck** [comment]
@luigiw  I think it would be very useful with an example in https://github.com/Azure/azureml-examples or similar. :-) Does it work by always setting PIP_INDEX_URL when the environment is building, or does it do some matching with url specified with `--index-url` in requirement files? The first would be cool as this means that the environment is not dependent on the the index keeping the specific url.

**Comment by luigiw** [comment]
I don't think you need to do anything special once a workspace connection is added for your private feed. @vizhur correct me if I was wrong here.

**Comment by vizhur** [comment]
for base image + conda spec case we'll update urls from the spec with the auth from ws connections using the closest match (if there are multiple connections for the same host name)

**Comment by thomasfrederikhoeck** [comment]
@vizhur   Hmm. okay. This issue was actually created because this didn't happen with managed endpoints but only for v1 ACIs? 
I don't think the logic surrounding private pip is very transparent. What happens if I create a private pip connection but don't specify a `--index-url` in the `pip` section of the conda spec? Is it still used?

**Comment by thomasfrederikhoeck** [comment]
@luigiw @vizhur What you describe is actually what I did (see top of Issue) but it failed? :-) It only looks like updating of URLs happen in environments created by the user and **not** the ones created automatically by AzureML when deploying a MLFlow model to managed endpoints (where the env can't be controlled).

**Comment by santiagxf** [comment]
@thomasfrederikhoeck this feature is now in private preview in AzureML using a capability called Packages. You can test it out by signing up here: [https://aka.ms/AzureML/ModelPackages ](https://aka.ms/AzureML/ModelPackages)
